File: android/src/main/java/com/reactnativemsal/FileUtils.java
package com.reactnativemsal;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * Code copied from org.apache.commons.io/FileUtils.java and org.apache.commons.io/IOUtils.java,
 * v2.8
 */
public class FileUtils {
    public static final int DEFAULT_BUFFER_SIZE = 8192;
    public static final int EOF = -1;

    public static void copyInputStreamToFile(InputStream source, File destination)
            throws IOException {
        try (InputStream inputStream = source) {
            copyToFile(inputStream, destination);
        }
    }

    public static void copyToFile(InputStream inputStream, File file) throws IOException {
        try (OutputStream out = openOutputStream(file, false)) {
            copy(inputStream, out);
        }
    }

    public static FileOutputStream openOutputStream(File file, boolean append) throws IOException {
        if (file == null) {
            throw new IOException("File cannot be null.");
        }
        if (file.exists()) {
            if (!file.isFile()) {
                throw new IOException("Not a file: " + file);
            }
            if (!file.canWrite()) {
                throw new IOException("File is not writable: '" + file + "'");
            }
        } else {
            File parent = file.getParentFile();
            if (parent != null) {
                if (!parent.mkdirs() && !parent.isDirectory()) {
                    throw new IOException("Cannot create directory '" + parent + "'.");
                }
            }
        }
        return new FileOutputStream(file, append);
    }

    public static int copy(InputStream inputStream, OutputStream outputStream) throws IOException {
        long count = copyLarge(inputStream, outputStream);
        if (count > Integer.MAX_VALUE) {
            return EOF;
        }
        return (int) count;
    }

    public static long copyLarge(InputStream inputStream, OutputStream outputStream)
            throws IOException {
        return copy(inputStream, outputStream, DEFAULT_BUFFER_SIZE);
    }

    public static long copy(InputStream inputStream, OutputStream outputStream, int bufferSize)
            throws IOException {
        return copyLarge(inputStream, outputStream, new byte[bufferSize]);
    }

    public static long copyLarge(InputStream inputStream, OutputStream outputStream, byte[] buffer)
            throws IOException {
        long count = 0;
        int n;
        while (EOF != (n = inputStream.read(buffer))) {
            outputStream.write(buffer, 0, n);
            count += n;
        }
        return count;
    }
}

File: android/src/main/java/com/reactnativemsal/FileUtils.kt
package com.reactnativemsal

import java.io.*

/**
 * Code copied from org.apache.commons.io/FileUtils.java and org.apache.commons.io/IOUtils.java, v2.6
 * and converted to Kotlin via Android Studio tools.
 */
object FileUtils {
  private const val DEFAULT_BUFFER_SIZE = 1024 * 4
  const val EOF = -1

  @Throws(IOException::class)
  fun copyInputStreamToFile(source: InputStream, destination: File) {
    source.use { `in` -> copyToFile(`in`, destination) }
  }

  @Throws(IOException::class)
  fun copyToFile(source: InputStream, destination: File) {
    source.use { `in` -> openOutputStream(destination).use { out -> copy(`in`, out) } }
  }

  @JvmOverloads
  @Throws(IOException::class)
  fun openOutputStream(file: File, append: Boolean = false): FileOutputStream {
    if (file.exists()) {
      if (file.isDirectory) {
        throw IOException("File '$file' exists but is a directory")
      }
      if (!file.canWrite()) {
        throw IOException("File '$file' cannot be written to")
      }
    } else {
      val parent = file.parentFile
      if (parent != null) {
        if (!parent.mkdirs() && !parent.isDirectory) {
          throw IOException("Directory '$parent' could not be created")
        }
      }
    }
    return FileOutputStream(file, append)
  }

  @Throws(IOException::class)
  fun copy(input: InputStream, output: OutputStream): Int {
    val count = copyLarge(input, output)
    return if (count > Int.MAX_VALUE) {
      -1
    } else count.toInt()
  }

  @Throws(IOException::class)
  fun copyLarge(input: InputStream, output: OutputStream): Long {
    return copy(input, output, DEFAULT_BUFFER_SIZE)
  }

  @Throws(IOException::class)
  fun copy(input: InputStream, output: OutputStream, bufferSize: Int): Long {
    return copyLarge(input, output, ByteArray(bufferSize))
  }

  @Throws(IOException::class)
  fun copyLarge(input: InputStream, output: OutputStream, buffer: ByteArray?): Long {
    var count: Long = 0
    var n: Int
    while (EOF != input.read(buffer).also { n = it }) {
      output.write(buffer, 0, n)
      count += n.toLong()
    }
    return count
  }
}
